### PR TITLE
initial implementation of ViomiZoneCleaningCapability class

### PR DESCRIPTION
**Pull request by rromrrom**
now user is able to edit zones. Zone cleaning does not works at the moment yet.

### CODE REVIEW

**Comment by Hypfer** [comment]
> Zone cleaning does not works at the moment yet.

Uhm. Is this ready to merge then?

**Comment by rromrrom** [comment]
Still testing. I will let you know when I'm done.

**Comment by Hypfer** [comment]
In that case, you can use the draft PR feature :)

File: lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+        valetudoZones.forEach( zone => {
+            const pA = ViomiMapParser.positionToViomi(zone.points.pA.x, zone.points.pA.y);
+            const pB = ViomiMapParser.positionToViomi(zone.points.pB.x, zone.points.pB.y);
+            const pC = ViomiMapParser.positionToViomi(zone.points.pC.x, zone.points.pC.y);
+            const pD = ViomiMapParser.positionToViomi(zone.points.pD.x, zone.points.pD.y);
+
+            for (let j = 0; j < zone.iterations; j++) {
+                areas.push([areas.length,
+                    0,
```
Comment (rumpeltux) [inline lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js:21]: use named constants. or at least document the meaning: 0: regular zone, 2: restricted area

File: lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+                    pC.x.toFixed(4),
+                    pC.y.toFixed(4),
+                    pD.x.toFixed(4),
+                    pD.y.toFixed(4)
+                ].join("_"));
+            }
+        });
+
+        Logger.trace("areas to clean: ", areas);
+        await this.robot.sendCommand("set_zone", [areas.length].concat(areas), {}).finally(() => {
```
Comment (rumpeltux) [inline lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js:35]: Wouldn’t this also need to issue a start command?

See the previous implementation here:
https://github.com/rumpeltux/Valetudo/blob/d5d9ec2fc78a6d290ce0be831a5ffc7c41da8a33/lib/devices/Viomi.js#L589

File: lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+            this.robot.pollMap();
```
Comment (rumpeltux) [inline lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js:36]: Can you document why the map needs to be pulled after setting the zone?
I don’t recall exactly, is the currently set zone populated in the map?

File: lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js (hunk 1 above)
Comment (rromrrom) [reply lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js]: done

File: lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js (hunk 2 above)
Comment (rromrrom) [reply lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js]: done

File: lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js (hunk 3 above)
Comment (rromrrom) [reply lib/robots/viomi/capabilities/ViomiZoneCleaningCapability.js]: got this from old implementation, skipped now

**Review by rumpeltux — APPROVED** [review]

**Pull request merged**